addons/stock/models/stock_scrap.py

```diff
@@ -67,49 +58,38 @@ def _compute_product_uom_id(self):
         for scrap in self:
             scrap.product_uom_id = scrap.product_id.uom_id
 
+    @api.depends('company_id', 'picking_id')
+    def _compute_location_id(self):
+        groups = self.env['stock.warehouse']._read_group(
+            [('company_id', 'in', self.company_id.ids)], ['min_id:min(id)'], ['company_id'])
+        locations_per_company = {
+            group['company_id'][0]: self.env['stock.warehouse'].browse(group['min_id']).lot_stock_id
+            for group in groups
+        }
+        for scrap in self:
+            if scrap.picking_id:
+                scrap.location_id = scrap.picking_id.location_dest_id if scrap.picking_id.state == 'done' else scrap.picking_id.location_id
+            else:
+                scrap.location_id = locations_per_company[scrap.company_id.id]
+
+    @api.depends('company_id')
+    def _compute_scrap_location_id(self):
+        groups = self.env['stock.location']._read_group(
+            [('company_id', 'in', self.company_id.ids), ('scrap_location', '=', True)], ['min_id:min(id)'], ['company_id'])
+        locations_per_company = {
+            group['company_id'][0]: self.env['stock.location'].browse(group['min_id'])
+            for group in groups
+        }
+        for scrap in self:
+            scrap.scrap_location_id = locations_per_company[scrap.company_id.id]
+
     @api.depends('move_id', 'move_id.move_line_ids.qty_done')
     def _compute_scrap_qty(self):
         self.scrap_qty = 1
         for scrap in self:
             if scrap.move_id:
                 scrap.scrap_qty = scrap.move_id.quantity_done
 
-    @api.onchange('picking_id')
-    def _onchange_picking_id(self):
-        if self.picking_id:
-            self.location_id = (self.picking_id.state == 'done') and self.picking_id.location_dest_id.id or self.picking_id.location_id.id
-
-    @api.onchange('product_id')
-    def _onchange_product_id(self):
-        if self.product_id:
-            if self.tracking == 'serial':
-                self.scrap_qty = 1
```

Comment:
You'll need to include this condition in `_compute_scrap_qty` because the scrap form view that opens via a picking's "Scrap" button makes the scrap quantity readonly when the product selected is serial (strange that the scrap form view via "Operations > Scrap > Create" flow doesn't :thinking: but probably OK for now I guess, especially if you change it so it switches to 1 whenever the product is changed). A demo of why this can be an issue is: 
[Odoo - WH_IN_00012.webm](https://user-images.githubusercontent.com/57942143/203553286-41b46652-989d-41b3-87af-fde80e522cbc.webm)

Reply:
adapted depends of _compute_scrap_qty and slightly refactored stock_scrap_form_view & stock_scrap_form_view2